---
title: "2020-42-set2_hto_souporcell_w_new_run"
author: "Timothée B"
date: "22/02/2022"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: hide
  pdf_document: 
    toc: true
    toc_depth: 3
    highlight: zenburn
editor_options: 
  chunk_output_type: inline
---
---
title:  "Seurat analysis on 2020-42-set2 data"
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = "/home/bigott/workspace2/01_tim/01_these/01_etv6/06_Analysis/")
```

## Loading Libraries 
```{r env_loading, include=FALSE}
# Load all necessary packages
library(sctransform)
library(reticulate)
library(Seurat)
library(htmlwidgets)
library(DT)
library(plotly)
library(pheatmap)
library(reshape2)
library("gridExtra")
library(knitr)
library(scater)
```

## Setting PATH
```{r}
setwd("/home/bigott/workspace2/01_tim/01_these/01_etv6/05_Outputs/02_2020-42-set2/")
# Path to the folder containing scripts used in the analysis
CWD <- "/home/bigott/workspace2/01_tim/01_these/01_etv6/03_scripts/" 
#  Path to the folder that will contain output objects
OUTPUT_PATH <- "/home/bigott/workspace2/01_tim/01_these/01_etv6/05_Outputs/06_Février_2022/02_2020-42-set2/"

# Load the R scripts containing the functions used in the analysis
source(paste(CWD, "01_general/Mito_function.R", sep="/"))

# Set the random number seed
set.seed(1234)

# Resolution parameter for Seurat clustering
RESOLUTION <- 1

# Load path for files
PATH_HUMAN_DATA1 <- "/home/bigott/workspace2/01_tim/01_these/01_etv6/04_data/02_2020-42-set2/01_cellranger_run2020-42-set2_GRCh382020A_CR5/outs/filtered_feature_bc_matrix/"
SAMPLE1 <- "ETV6_D11"
PATH_HTO_DATA1 <- "/home/bigott/workspace2/01_tim/01_these/01_etv6/04_data/02_2020-42-set2/03_run_CITEseqcount_26_02_22/umi_count/"
PATH_SOUPORCELL_DATA1 <- "/home/bigott/workspace2/01_tim/01_these/02_Souporcell/02_2020-42-set2/02_run_26_02_22/"
```

## Data preprocessing {.tabset}
```{r Sample_loading, include=FALSE}
# Read 10X data
etv6_data <- Read10X(data.dir = PATH_HUMAN_DATA1)

# Create the Seurat object and applies filters 
Not_processed_Seurat1 <- CreateSeuratObject(counts = etv6_data, min.cells = 3, min.features = 200, project = "ETV6")
```

```{r HTO1_loading, echo=TRUE, message=FALSE}
# Load in the UMI matrix
umi_sparse1 <- GetAssayData(object = Not_processed_Seurat1, slot = "counts")

#je retire les -1 dans le nom des barcodes:
colnames(umi_sparse1)=(substr(colnames(umi_sparse1),1,16))

# For generating a hashtag count matrix from fastq files, please refer to https://github.com/Hoohm/CITE-seq-Count.
# Load in the HTO count matrix
raw.hto1 <- Read10X(PATH_HTO_DATA1, gene.column = 1)
hto1 <- raw.hto1[c(1:4),]


rownames(hto1) <- c("F417.II.2","P214.III.8","Ctrl1","Ctrl2")

# Select cell barcodes detected by both RNA and HTO
# In the example datasets we have already filtered the cells for you, but perform this step for clarity.
joint_bcs1 <- intersect(colnames(umi_sparse1),colnames(hto1))

# Subset RNA and HTO counts by joint cell barcodesumi_sparse <- pbmc_umi_sparse[,joint_bcs]
hto1 <- as.matrix(hto1[,joint_bcs1])
filtered.raw.hto1 <- as.matrix(raw.hto1[,joint_bcs1])

# Confirm that the HTO have the correct names
print (rownames(hto1))
# [1] "F417.II.2"  "P214.III.8" "Ctrl1"      "Ctrl2"
```

```{r echo=TRUE}
par(mfrow=c(1,1))
par(las=2)
par(mar=c(5,15,3,3))
barplot(rowSums(hto1), main = "sequenced HTO distribution", horiz=TRUE)

rowSums(hto1)

htotest <- rbind(hto1,colSums(hto1))
rownames(htotest) <- c("F417.II.2","P214.III.8","Ctrl1","Ctrl2","sum")

htotest <- t(htotest)
htotest <- as.data.frame(htotest)
```

```{r echo=TRUE, message=FALSE}
### Setup seurat object and add in the hto data
# Setup Seurat object
hashtag1 <- CreateSeuratObject(counts = umi_sparse1[,joint_bcs1], assay = "RNA")

# Normalize RNA data with log normalization
hashtag1 <- NormalizeData(hashtag1,display.progress = F)
# Find and scale variable genes
hashtag1 <- FindVariableFeatures(hashtag1, do.plot = F, selection.method = "vst", nfeatures = 2000, display.progress = F)
hashtag1 <- ScaleData(hashtag1,genes.use = rownames(hashtag1)) #, vars.to.regress = c("S.Score", "G2M.Score") ,display.progress = FALSE)
```

```{r, message=FALSE, include=FALSE}
### Adding HTO data as an independent assay

# Add HTO data as a new assay independent from RNA
hashtag1[["HTO"]] <- CreateAssayObject(counts = hto1)
hashtag1 <- SetAssayData(hashtag1,assay = "HTO",slot = "counts",new.data = hto1)
# Normalize HTO data, here we use centered log-ratio (CLR) transformation
hashtag1 <- NormalizeData(hashtag1, assay = "HTO",normalization.method = "CLR",display.progress = FALSE)
```

```{r, message=FALSE, include=FALSE}
#Demultiplex cells based on HTO enrichment

#Here we use the Seurat function HTODemux() to assign single cells back to their sample origins.

# If you have a very large dataset we suggest using k_function = "clara". This is a k-medoid clustering function for large applications
# You can also play with additional parameters (see documentation for HTODemux()) to adjust the threshold for classification
# Here we are using the default settings
hashtag1 <- HTODemux(hashtag1, assay = "HTO", positive.quantile = 0.99, verbose = FALSE)
hashtag1 <- MULTIseqDemux(hashtag1, assay = "HTO",autoThresh = TRUE, maxiter = 10,qrange = seq(from = 0.1, to = 0.9, by = 0.05), verbose = TRUE)
```

```{r, message=FALSE}
################
### Hash.ID ####
################

par(mfrow=c(2,3))
par(las=2)
par(mar=c(8,3,1,1))

HTO_1 <- row.names(subset(hashtag1@meta.data, hash.ID == "F417.II.2" ))
boxplot(t(hto1[,HTO_1]), ylim = c(0,1000))
title(main = "F417.II.2 (hash.ID)")
abline(h = 100)

HTO_6 <- row.names(subset(hashtag1@meta.data, hash.ID == "P214.III.8" ))
boxplot(t(hto1[,HTO_6]), ylim = c(0,1000))
title(main = "P214.III.8 (hash.ID)")
abline(h = 100)

HTO_7 <- row.names(subset(hashtag1@meta.data, hash.ID == "Ctrl1" ))
boxplot(t(hto1[,HTO_7]), ylim = c(0,1000))
title(main = "Ctrl1 (hash.ID)")
abline(h = 100)

HTO_8 <- row.names(subset(hashtag1@meta.data, hash.ID == "Ctrl2" ))
boxplot(t(hto1[,HTO_8]), ylim = c(0,1000))
title(main = "Ctrl2 (hash.ID)")
abline(h = 100)

##################
#### MULTI_ID ####
##################

par(mfrow=c(2,3))
par(las=2)
par(mar=c(8,3,1,1))

HTO_1 <- row.names(subset(hashtag1@meta.data, MULTI_ID == "F417.II.2" ))
boxplot(t(hto1[,HTO_1]), ylim = c(0,1000))
title(main = "F417.II.2 (Multi_ID)")
abline(h = 100)

HTO_6 <- row.names(subset(hashtag1@meta.data, MULTI_ID == "P214.III.8" ))
boxplot(t(hto1[,HTO_6]), ylim = c(0,1000))
title(main = "P214.III.8 (Multi_ID)")
abline(h = 100)

# HTO_7 <- row.names(subset(hashtag1@meta.data, MULTI_ID == "Ctrl1" ))
# boxplot(t(hto1[,HTO_7]), ylim = c(0,1000))
# title(main = "Ctrl1 (Multi_ID)")
# abline(h = 100)

HTO_8 <- row.names(subset(hashtag1@meta.data, MULTI_ID == "Ctrl2" ))
boxplot(t(hto1[,HTO_8]), ylim = c(0,1000))
title(main = "Ctrl2 (Multi_ID)")
abline(h = 100)
```

## Demultiplexing results {.tabset}

## Cells classification

Cells classification as singlets, doublets and negative/ambiguous cells for HTOdemux and 
doublet, negative and HTO for Multiseq
Overall, multiseq results seem  better than HTOdemux

- HTOdemux

```{r, message=FALSE}
# Global classification results
print (table(hashtag1@meta.data$HTO_classification.global))
print (table(hashtag1@meta.data$HTO_classification))
```

- Multiseq 

```{r, message=FALSE}
print (table(hashtag1@meta.data$MULTI_ID))
```
  
## Scatterplot

- HTOdemux

```{r echo=TRUE, message=FALSE}
FeatureScatter(hashtag1, feature1 = "F417.II.2", feature2 = "P214.III.8", group.by = "hash.ID")
FeatureScatter(hashtag1, feature1 = "F417.II.2", feature2 = "Ctrl1", group.by = "hash.ID")
FeatureScatter(hashtag1, feature1 = "F417.II.2", feature2 = "Ctrl2", group.by = "hash.ID")

FeatureScatter(hashtag1, feature1 = "P214.III.8", feature2 = "Ctrl1", group.by = "hash.ID")
FeatureScatter(hashtag1, feature1 = "P214.III.8", feature2 = "Ctrl2", group.by = "hash.ID")

FeatureScatter(hashtag1, feature1 = "Ctrl1", feature2 = "Ctrl2", group.by = "hash.ID")
```

- Multiseq 

```{r echo=TRUE, message=FALSE}
FeatureScatter(hashtag1, feature1 = "F417.II.2", feature2 = "P214.III.8", group.by = "MULTI_ID")
FeatureScatter(hashtag1, feature1 = "F417.II.2", feature2 = "Ctrl1", group.by = "MULTI_ID")
FeatureScatter(hashtag1, feature1 = "F417.II.2", feature2 = "Ctrl2", group.by = "MULTI_ID")

FeatureScatter(hashtag1, feature1 = "P214.III.8", feature2 = "Ctrl1", group.by = "MULTI_ID")
FeatureScatter(hashtag1, feature1 = "P214.III.8", feature2 = "Ctrl2", group.by = "MULTI_ID")

FeatureScatter(hashtag1, feature1 = "Ctrl1", feature2 = "Ctrl2", group.by = "MULTI_ID")
```
   
## Violinplot (features)

- HTOdemux

```{r, message=FALSE}
VlnPlot(hashtag1,features = c("nFeature_RNA"),pt.size = 0.1, log = TRUE,  group.by = "hash.ID")
```

- Multiseq 

```{r, message=FALSE}
VlnPlot(hashtag1,features = c("nFeature_RNA"),pt.size = 0.1, log = TRUE,  group.by = "MULTI_ID")
```

## Violin plots (HTO counts)

- HTOdemux

```{r, message=FALSE}
VlnPlot(hashtag1,features = c("nCount_HTO"),pt.size = 0.1, log = TRUE,  group.by = "hash.ID")
```

- Multiseq 

```{r, message=FALSE}
VlnPlot(hashtag1,features = c("nCount_HTO"),pt.size = 0.1, log = TRUE,  group.by = "MULTI_ID")
```

```{r, message=FALSE}
# Calculate a distance matrix using HTO
hto.dist.mtx <- as.matrix(dist(t(GetAssayData(object = hashtag1, assay = "HTO"))))

# Calculate tSNE embeddings with a distance matrix
hashtag1 <- RunTSNE(hashtag1, distance.matrix = hto.dist.mtx, perplexity = 100)
```

## UMAP based on HTO

- HTOdemux

```{r, message=FALSE}
DimPlot(hashtag1, group.by = "hash.ID")
```

- Multiseq 

```{r, message=FALSE}
DimPlot(hashtag1, group.by = "MULTI_ID")
```

- HTO_maxID

```{r, message=FALSE}
DimPlot(hashtag1, group.by = "HTO_maxID")
```

- HTO margin

```{r echo=TRUE, fig.height=7, fig.width=8, message=FALSE}
Tsne<-data.frame(
  tSNE_1 = hashtag1@reductions$tsne@cell.embeddings[,1],
  tSNE_2= hashtag1@reductions$tsne@cell.embeddings[,2],
  gene= hashtag1@meta.data$HTO_margin
)

HTO= hashtag1@meta.data$MULTI_ID
Max=max(hashtag1@meta.data$HTO_margin)
Min=min(hashtag1@meta.data$HTO_margin)

ggplotly(ggplot(Tsne,aes(x=tSNE_1,y=tSNE_2))+geom_point(aes(color=gene,shape=HTO))+
    scale_colour_gradient2(low = "blue",mid="orange",high="red",name="HTO margin",midpoint=(Max+Min)/2)+scale_shape_manual(values = c(15,16,17,18,19,20,21,22,23,24,25)),tooltip = "all")
```

## Ridge plots

**Visualize enrichment for selected HTOs with ridge plots**

- Group cells based on the max HTO signal (HTOdemux)

```{r, fig.height = 4, message=FALSE}
RidgePlot(hashtag1, assay = "HTO", features = rownames(GetAssayData(hashtag1,assay = "HTO"))[1:2],ncol = 2, group.by = "HTO_maxID")
RidgePlot(hashtag1, assay = "HTO", features = rownames(GetAssayData(hashtag1,assay = "HTO"))[3:4],ncol = 2, group.by = "HTO_maxID")

RidgePlot(hashtag1, assay = "HTO", features = rownames(GetAssayData(hashtag1,assay = "HTO"))[1:2],ncol = 2, group.by = "hash.ID")
RidgePlot(hashtag1, assay = "HTO", features = rownames(GetAssayData(hashtag1,assay = "HTO"))[3:4],ncol = 2, group.by = "hash.ID")
```

- Group cells based on the max HTO signal (Multiseq)

```{r, fig.height = 4, message=FALSE}
RidgePlot(hashtag1, assay = "HTO", features = rownames(GetAssayData(hashtag1,assay = "HTO"))[1:2],ncol = 2, group.by = "MULTI_ID")
RidgePlot(hashtag1, assay = "HTO", features = rownames(GetAssayData(hashtag1,assay = "HTO"))[3:4],ncol = 2, group.by = "MULTI_ID")
```

## Souporcell 
```{r}
souporcell_clusters <- read.table(file = paste(PATH_SOUPORCELL_DATA1, "clusters.tsv", sep ="/"), sep= "\t", header = T)
table(souporcell_clusters$status)
# je retire les -1 dans le nom des barcodes:
souporcell_clusters$barcode=(substr(souporcell_clusters$barcode,1,16))
hashtag1@meta.data$barcode <- rownames(hashtag1@meta.data) 

# je merge les datasets en fontions du nombre de cellules dans le dataset seurat not process + HTO
joint_souporcell_hto <- left_join(hashtag1@meta.data,souporcell_clusters)

rownames(joint_souporcell_hto) <- joint_souporcell_hto$barcode
# hashtag1@meta.data <- joint_souporcell_hto

##############################################################################
# Check Souporcell clusters identity
# You can attribute an identity if there are a lot of occurences of 1 sample.
# I also check if HTO demux found a singlet to be sure.
##############################################################################

# Ctrl1
# table(joint_souporcell_hto$HTO_classification[which(joint_souporcell_hto$assignment=="0")])
# table(joint_souporcell_hto$HTO_maxID[which(joint_souporcell_hto$assignment=="0")])
# P214.III.8
# table(joint_souporcell_hto$HTO_classification[which(joint_souporcell_hto$assignment=="1")])
# table(joint_souporcell_hto$HTO_maxID[which(joint_souporcell_hto$assignment=="1")])
# Ctrl2
# table(joint_souporcell_hto$HTO_classification[which(joint_souporcell_hto$assignment=="2")])
# table(joint_souporcell_hto$HTO_maxID[which(joint_souporcell_hto$assignment=="2")])
# F417.II.2
# table(joint_souporcell_hto$HTO_classification[which(joint_souporcell_hto$assignment=="3")])
# table(joint_souporcell_hto$HTO_maxID[which(joint_souporcell_hto$assignment=="3")])

# Identify cells not present in souporcell file
joint_souporcell_hto[which(is.na(joint_souporcell_hto$status) == TRUE),]
# "AGCGTCGGTTCAGGCC" "ATTGGACGTCGTTGTA" "CGATGTAGTTGTGGCC" "CGGTTAATCTGTACGA" "CGTAGGCTCAGAAATG" "GCAGTTATCAGAGGTG" "GCTTGAAAGCTAGTGG" "TCGAGGCGTGTTTGTG"

# Remove cells not present in souporcell file
# joint_souporcell_hto<- joint_souporcell_hto[-which(is.na(joint_souporcell_hto$status) == TRUE),]

#############################################################################
# We only keep souporcell singlet by cluster 
#############################################################################

joint_souporcell_hto$souporcell_classif <- "NA"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$status == "doublet",]$barcode,"-1",""),]$souporcell_classif <- "doublet"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$status == "unassigned",]$barcode,"-1",""),]$souporcell_classif <- "unassigned"

joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$assignment == "0" & joint_souporcell_hto$status == "singlet",]$barcode,"-1",""),]$souporcell_classif <- "cluster0"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$assignment == "1" & joint_souporcell_hto$status == "singlet",]$barcode,"-1",""),]$souporcell_classif <- "cluster1"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$assignment == "2" & joint_souporcell_hto$status == "singlet",]$barcode,"-1",""),]$souporcell_classif <- "cluster2"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$assignment == "3" & joint_souporcell_hto$status == "singlet",]$barcode,"-1",""),]$souporcell_classif <- "cluster3"

joint_souporcell_hto$HT0_souporcell_classif <- paste(joint_souporcell_hto$HTO_classification.global, joint_souporcell_hto$souporcell_classif, sep="-")
table(joint_souporcell_hto$HT0_souporcell_classif)

#######################################################################################################
# We keep cells identify as a Singlet in at least one of the tools (HTO demux or Souporcell)
# True doublet are doublet identify in both analysis and unassigned are Negative and/or doublet in both.
#######################################################################################################

joint_souporcell_hto$HT0_souporcell_classif_final <- "NA"

joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Doublet-doublet",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "doublet"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Doublet-unassigned",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "unassigned"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Negative-doublet",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "unassigned"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Negative-unassigned",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "unassigned"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Singlet-doublet",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "unassigned"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Singlet-unassigned",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "unassigned"

joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Doublet-cluster0",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster0"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Doublet-cluster1",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster1"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Doublet-cluster2",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster2"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Doublet-cluster3",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster3"


joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Negative-cluster0",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster0"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Negative-cluster1",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster1"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Negative-cluster2",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster2"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Negative-cluster3",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "cluster3"


joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Singlet-cluster0",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "Singlet"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Singlet-cluster1",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "Singlet"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Singlet-cluster2",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "Singlet"
joint_souporcell_hto[gsub(x = joint_souporcell_hto[joint_souporcell_hto$HT0_souporcell_classif == "Singlet-cluster3",]$barcode,"-1",""),]$HT0_souporcell_classif_final <- "Singlet"

table(joint_souporcell_hto$HT0_souporcell_classif_final)

```

## Select cells
```{r results='asis'}
cluster0 <- row.names(subset(joint_souporcell_hto, HT0_souporcell_classif_final == "cluster0" ))
cluster1 <- row.names(subset(joint_souporcell_hto, HT0_souporcell_classif_final == "cluster1" ))
cluster2 <- row.names(subset(joint_souporcell_hto, HT0_souporcell_classif_final == "cluster2" ))
cluster3 <- row.names(subset(joint_souporcell_hto, HT0_souporcell_classif_final == "cluster3" ))
Singlet <- row.names(subset(joint_souporcell_hto, HT0_souporcell_classif_final == "Singlet" ))

##########################################################################################################
# We define an identity for each souporcell cluster, based on the maximum of occurences for a given sample
# Indeed doublet or undefined HTO identify cells don't have a real identity for this tool. 
# Souporcell give us the opportunity to integrate these cells in the dataset.
##########################################################################################################

Clusters_identified1 = c(cluster0,cluster1,cluster2,cluster3,Singlet)
hashtag1@meta.data <- joint_souporcell_hto
clean.subset1 <- subset(x = hashtag1, cells = Clusters_identified1)
VlnPlot(clean.subset1,features = c("nFeature_RNA", "nCount_RNA"),pt.size = 0.1, log = TRUE,  group.by = "HT0_souporcell_classif_final")
table(clean.subset1$HT0_souporcell_classif_final)

for (i in 1:length(rownames(clean.subset1@meta.data))){
  if (clean.subset1$HT0_souporcell_classif_final[i] == "cluster0"){
    clean.subset1@meta.data$HT0_souporcell_classification[i] <- "Ctrl1"
  }
  if (clean.subset1$HT0_souporcell_classif_final[i] == "cluster1"){
    clean.subset1@meta.data$HT0_souporcell_classification[i] <- "P214.III.8"
  }
  if (clean.subset1$HT0_souporcell_classif_final[i] == "cluster2"){
    clean.subset1@meta.data$HT0_souporcell_classification[i] <- "F417.II.2"
  }
  if (clean.subset1$HT0_souporcell_classif_final[i] == "cluster3"){
    clean.subset1@meta.data$HT0_souporcell_classification[i] <- "Ctrl2"
  }
  if (clean.subset1$HT0_souporcell_classif_final[i] == "Singlet"){
    clean.subset1@meta.data$HT0_souporcell_classification[i] <- clean.subset1$HTO_classification[i]
  }
}
# quantif <-  paste(test$HT0_souporcell_classif_final,test$HTO_classification, sep="_")
# table(quantif)
table(clean.subset1@meta.data$HT0_souporcell_classification)
```

```{r, message=FALSE}
DimPlot(clean.subset1, group.by = "HT0_souporcell_classif_final")
DimPlot(clean.subset1, group.by = "HT0_souporcell_classification")
```

## ALL samples together analysis

## Percentage of mitochondrial genes
```{r include=FALSE}

mito.genes <- grep("MT-", rownames(clean.subset1@assays$RNA), value=T)
mito.genes

# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
clean.subset1[["percent.mt"]] <- PercentageFeatureSet(clean.subset1, pattern = "MT-")

# Compute mitochondrial percentage threshold
discard.mito=isOutlier(clean.subset1[["percent.mt"]][,1],type="higher")
mito.threshold=min(clean.subset1[["percent.mt"]][,1][discard.mito])
```

## Visualize QC metrics as a violin plot
```{r}
VlnPlot(clean.subset1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3,pt.size=0.1, group.by = "HT0_souporcell_classification")
```

## Subset the dataset, to remove the cells having too few genes/cell and too much mitochondrial genes. Here, I'll set the cut-offs at min 1000 genes and max percent mitochondrial genes `r signif(mito.threshold,3)`% (calculated by function scater::isOutlier()).  
```{r}
clean.subset1 <- subset(clean.subset1, subset = (nFeature_RNA > 1000 & percent.mt < mito.threshold))
dim(clean.subset1)
```

## Normalize data
```{r}
# standard log-normalization
etv6_clean <- NormalizeData(clean.subset1, normalization.method = "LogNormalize", scale.factor = 10000)

# write.table(etv6_clean@assays$RNA@data,
# 	    paste("log_norm_filtered_data_etv6_clean.txt",sep=""),
# 	    sep="\t", quote=F, col.names=NA)
```

## Some counts after normalization
```{r}
etv6_clean
table(etv6_clean$HT0_souporcell_classification)
```

## Choose top 1k variable genes
```{r}
etv6_clean <- FindVariableFeatures(object=etv6_clean, selection.method = "vst", mean.function = ExpMean, dispersion.function = LogVMR, binning.method = "equal_width", num.bin = 20, y.cutoff = 0.5, nfeatures = 1000, verbose=TRUE)
#write.table(icare@assays$RNA@var.features,
#	     paste("Top1000_variable_genes.",experiment,"_",params$sample,".txt", sep=""),
#	     sep="\t", quote=F, col.names=NA
#)

top40 <- head(VariableFeatures(etv6_clean), 40)
top100 <- head(VariableFeatures(etv6_clean), 100)
```

## Plot variable genes:
```{r}
plot1 <- VariableFeaturePlot(etv6_clean)
LabelPoints(plot = plot1, points = top40, repel = TRUE)
```

## Ggplot function
```{r}
get_cluster_centroids <- function(comp_1,comp_2) {
	coords=cbind(comp_1,comp_2)
	clusters=clusters_res
	centers=c()
	for (cl in unique(clusters)) {
		sel=clusters==cl
		x_center=mean(coords[sel,1])
		y_center=mean(coords[sel,2])
		centers=rbind(centers,c(cl,x_center,y_center))
	}
	centers=data.frame("Cluster"=as.factor(centers[,1]),
			   "x_center"=as.numeric(centers[,2]),
			   "y_center"=as.numeric(centers[,3])
			   )
}

color_ordered2=c("indianred1", "darkorange3", "olivedrab3", "red4", "palegreen", "royalblue4", "palevioletred3", "orchid3", "olivedrab2", "palegreen3", "lightseagreen", "steelblue2", "royalblue2", "turquoise1", "chocolate1", "firebrick1", "magenta", "mediumorchid1","darkgreen", "indianred2", "blue")
```

## Standard scaling (no regression)
```{r}
setwd(OUTPUT_PATH)
all.genes <- rownames(etv6_clean)
etv6_clean <- ScaleData(etv6_clean, features = all.genes, display.progress = TRUE, do.scale=FALSE)
# write.table(etv6_clean@assays$RNA@scale.data,
#     "log_norm_scaled_filtered_data_ECFC_all_samples.txt",
#     sep="\t", quote=F, col.names=NA
# )
```

## Run PCA
```{r}
etv6_clean <- RunPCA(etv6_clean, features = VariableFeatures(object = etv6_clean), npcs=40, pcs.print = 5, seed.use=42, rev.pca = FALSE)
```

## Examine PCA results
```{r}
Idents(etv6_clean) <- etv6_clean@meta.data$HT0_souporcell_classification
VizDimLoadings(etv6_clean, dims = 1:2, reduction = "pca")
DimPlot(etv6_clean, dims=c(1, 2), reduction = "pca", pt.size=1)
DimPlot(etv6_clean, dims=c(3, 4), reduction = "pca", pt.size=1)
```

## Save PCA coord
```{r}
pca_coordinates <- etv6_clean@reductions$pca@cell.embeddings
# write.table(pca_coordinates,
# 	    paste("PCA_coordinates_2020-42-set2_all_samples.txt", sep=""),
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Heatmaps:
```{r}
DimHeatmap(etv6_clean, dims = 1:6, cells = length(colnames(etv6_clean)), balanced = TRUE)
DimHeatmap(etv6_clean, dims = 1, cells = length(colnames(etv6_clean)), balanced = TRUE)
```

## JackStraw and ElbowPlot:
```{r}
etv6_clean <- JackStraw(etv6_clean, num.replicate = 100)
etv6_clean <- ScoreJackStraw(etv6_clean, dims = 1:20)
           
JackStrawPlot(etv6_clean, dims = 1:15)
           
ElbowPlot(etv6_clean)
```

## Perform clustering: 
```{r}
# setwd(OUTPUT_PATH)
etv6_clean <- FindNeighbors(object = etv6_clean, k.param = 30, dims = 1:6, compute.SNN = TRUE, prune.SNN = 1/15)

etv6_clean <- FindClusters(object= etv6_clean, modularity.fxn = 1,
             resolution = 0.8, algorithm = 1, n.start = 10, n.iter = 10,
             random.seed = 42, temp.file.location = NULL, edge.file.name = NULL,
             verbose = TRUE)
clusters_res <- etv6_clean[["RNA_snn_res.0.8"]][,1]
names(clusters_res)=colnames(etv6_clean)

# write.table(clusters_res,
# 	    "clusters_res_0.8_2020-42-set2_all_samples.txt",
# 	    sep="\t", quote=F, col.names=NA)
```

## Look at PCA results again, with clusters
```{r}
DimPlot(etv6_clean, dims=c(1, 2), reduction = "pca", pt.size=1)
DimPlot(etv6_clean, dims=c(3, 4), reduction = "pca", pt.size=1)
```

## Run UMAP: 
```{r}
# setwd(OUTPUT_PATH)
etv6_clean <- RunUMAP(etv6_clean, dims = 1:10, umap.method="uwot", seed.use=10, n.components=2, n.neighbors = 10, spread=1, min.dist=0.2)
DimPlot(etv6_clean, reduction = "umap", pt.size=0.7, label = TRUE)

# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
UMAP_coord <- etv6_clean@reductions$umap@cell.embeddings
# write.table(UMAP_coord,
# 	    "UMAP_coordinates_2020-42-set2_all_samples.txt",
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Run UMAP Cell cycling
```{r}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
 
etv6_clean <- CellCycleScoring(etv6_clean,s.features = s.genes, g2m.features = g2m.genes)
DimPlot(etv6_clean,group.by="Phase")+ggtitle("UMAP - Cell cycle")
```

## Run tSNE: 
```{r}
etv6_clean <- RunTSNE(object= etv6_clean,  dims = 1:10, dim.embed = 2, seed.use = 42, tsne.method="Rtsne", reduction="pca")
DimPlot(etv6_clean, reduction = "tsne", pt.size=0.7, label = TRUE)
tSNE_coord <- etv6_clean@reductions$tsne@cell.embeddings
```

## Top marker genes by clusters and exclude ribosomal genes
```{r}
# ribosomal_genes <- c("RPS17","RPS9","RPS9","RPS9","RPS9","RPS9","RPS9","RPS9","MRPS18B","MRPS18B","MRPS18B","RPS9","RPS18","MRPS36","RPS9","MRPS18B","RPS4Y2","RPS4Y1","RPS25","RPL7A","MRPS6","MRPS18B","RPS18","MRPS18B","RPS18","MRPL57","TBCE","RPS18","MRPS12","MRPS31","RPL3","MRPL23","RPL3L","RPS2","RPL17-C18orf32","RPS4X","RPS13","RPL36","RPS21","RPL21","RPL14","MRPS24","MRPL54","RPS24","RPS15A","RPS6","MRPS25","MRPL36","RPL36AL","MRPS17","MRPL16","MRPS30","FAU","MRPS16","RPS7","RPL23","RPS10","RPL10L","RPL35","RPL17","SRBD1","RPL22","RPL4","MRPS36","RPL37","MRPS2","RPL36A","RPL36A-HNRNPH2","RPL30","MRPL49","RPS19","RPL12","RPS5","RPS27L","RPS29","RPSA","MRPL33","MRPL1","RPS17","RPL10","MRPL13","MRPL19","MRPL15","MRPL41","RPS12","MRPS9","MRPS11","MRPL3","RPL31","RPS3","RPL38","RPL22L1","RPL39","MRPL46","RPL18A","RPL7","RPS3A","RPLP2","RPLP0","RPS16","MRPL23","MRPL22","MRPL18","RPL7L1","MRPL2","MRPS34","MRPS33","RPL28","RPL7A","MRPL4","RPL35A","RPL13","RPL24","RPS20","RPS15","MRPL14","MRPL21","MRPL32","NDUFA7","RPS28","RPL15","RPL18","RPL9","RPL37A","RPS11","MRPS23","MRPL52","MRPL43","MRPS18B","RPLP1","MRPS18A","RSL24D1","RPL27","RPL13A","MRPS14","MRPL47","MRPS35","RPL8","RPL26L1","MRPL11","RPL26","TBCE","RPS27AP5","MRPL34","MRPS12","MRPS5","MRPS18C","RPL29","RPL39L","MRPL28","MRPL51","RPL34","RPL10A","MRPL12","RPL32","RPS14","MRPL35","MRPL17","RPL11","RPS23","RPS9","RPL23A","RPS18","MRPL42","MRPL10","UBA52","DAP3","MRPS7","RPL19","RPS25","RPS26","RPL41","MRPS22","RPL6","RPS8","TBCE","MRPL30","RPS27","RPL27A","MRPS21","MRPL27","MRPL20","MRPL55","MRPL37","RPS27A","MRPL24","RPL5","MRPL9","MRPS15")
# 
# ribosomal_genes_filtered <- ribosomal_genes[!duplicated(ribosomal_genes)]
# all.genes_without_ribosomal = setdiff(all.genes, ribosomal_genes_filtered)
# # length(ribosomal_genes)
# # length(ribosomal_genes_filtered)
# # length(all.genes)
# # length(all.genes_without_ribosomal)
# for(i in 0:(max(as.numeric(levels(etv6_clean@meta.data$seurat_clusters))))) {
#   assign(paste("cluster", i,".markers", sep = ""), FindMarkers(etv6_clean, features = all.genes_without_ribosomal, ident.1 = i, test.use = "bimod", only.pos = TRUE))
#     print(paste("cluster", i,".markers", sep = ""))
#     print(head(get(paste("cluster", i,".markers", sep = "")),n=50))
#     write.table(get(paste("cluster", i,".markers", sep = "")),
# 		paste("cluster", i,"markers_2020-42-set2_all_samples.txt", sep = "_"),
# 		quote=FALSE,row.names = TRUE)
# }
```

## Dimensional reduction plot, with cells colored by a quantitative feature
```{r echo=FALSE}
gene_list <- c("TF","PROCR","CD31","PCAM1","CD34","THBD","ICAM1","MCAM","VCAM1","SELP","CD146","ENG","CDH5","KDR")


for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = etv6_clean, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 1))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}

```

## Metadata all samples
```{r}
# setwd(OUTPUT_PATH)
metadata <- as.data.frame(etv6_clean@meta.data)

#Substitute "RNA_snn_res.0.5" with "RNA_snn_res"
#colnames(metadata)=gsub("RNA_snn_res.[0-9.]+$","RNA_snn_res",colnames(metadata))
metadata$UMAP_1 <- etv6_clean@reductions$umap@cell.embeddings[,1]
metadata$UMAP_2 <- etv6_clean@reductions$umap@cell.embeddings[,2]
metadata$tSNE_1 <- etv6_clean@reductions$tsne@cell.embeddings[,1]
metadata$tSNE_2 <- etv6_clean@reductions$tsne@cell.embeddings[,2]
metadata$PC1 <- etv6_clean@reductions[["pca"]]@cell.embeddings[,1]
metadata$PC2 <- etv6_clean@reductions[["pca"]]@cell.embeddings[,2]

clusters_res <- etv6_clean[["RNA_snn_res.0.8"]][,1]
names(clusters_res)=colnames(etv6_clean)

# write.table(metadata,
# 	    "metadata_2020-42-set2_all_samples.txt",
# 	    sep='\t', quote=F, col.names=NA)

```

## Plot the gene/feature number per cell
```{r}
ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=nFeature_RNA), size=0.5)+ scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))
```

## Plot the umi/nCount number per cell
```{r}
ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=nCount_RNA), size=0.5)+ scale_colour_gradientn(colours=c("darkblue", "blue", "grey", "orange", "red")) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))
```

## Ggplot for all samples 
```{r}
centers=get_cluster_centroids(metadata$UMAP_1,metadata$UMAP_2)
ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.5) + scale_colour_manual(values=color_ordered2) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+
	geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,
		  #nudge_y=-1, nudge_x=-1,
		  size=3)

for(i in sort(unique(metadata$HT0_souporcell_classification))){
  metadata2 <- metadata[which(metadata$HT0_souporcell_classification == i),]
  clusters_res <- metadata2$RNA_snn_res.0.8
  p <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.5) + scale_colour_manual(values=color_ordered2) +       theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(i) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,
		  #nudge_y=-1, nudge_x=-1,
		  size=3)
  plot(p)
}
```

## Save the R object
```{r}
saveRDS(etv6_clean, file="2020-42-set2.rds")
```

## Subset on patients 
```{r}
Idents(clean.subset1) <- clean.subset1$HT0_souporcell_classification
clean.subset2 <- subset(x = clean.subset1, idents = c("P214.III.8", "F417.II.2"))
table(clean.subset2$HT0_souporcell_classification)
```

## Percentage of mitochondrial genes
```{r include=FALSE}

mito.genes <- grep("MT-", rownames(clean.subset2@assays$RNA), value=T)
mito.genes

# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
clean.subset2[["percent.mt"]] <- PercentageFeatureSet(clean.subset2, pattern = "MT-")

# Compute mitochondrial percentage threshold
discard.mito=isOutlier(clean.subset2[["percent.mt"]][,1],type="higher")
mito.threshold=min(clean.subset2[["percent.mt"]][,1][discard.mito])
```

## Filtering on total genes detected and cells with too much mitochondrial genes
```{r}
clean.subset2 <- subset(clean.subset2, subset = (nFeature_RNA > 1000 & percent.mt < mito.threshold))
dim(clean.subset2)
```

## Visualize QC metrics as a violin plot
```{r}
VlnPlot(clean.subset2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3,pt.size=0.1, group.by = "HT0_souporcell_classification")
```

## Normalize data
```{r}
# standard log-normalization
etv6_clean2 <- NormalizeData(clean.subset2, normalization.method = "LogNormalize", scale.factor = 10000)
# 
# write.table(etv6_clean2@assays$RNA@data,
# 	    "log_norm_filtered_2020-42-set2_patients.txt",
# 	    sep="\t", quote=F, col.names=NA)
```

## Some counts after normalization
```{r}
etv6_clean2
table(etv6_clean2$HT0_souporcell_classification)
```

## Choose top 1k variable genes
```{r}
etv6_clean2 <- FindVariableFeatures(object=etv6_clean2, selection.method = "vst", mean.function = ExpMean, dispersion.function = LogVMR, binning.method = "equal_width", num.bin = 20, y.cutoff = 0.5, nfeatures = 1000, verbose=TRUE)
#write.table(icare@assays$RNA@var.features,
#	     paste("Top1000_variable_genes.",experiment,"_",params$sample,".txt", sep=""),
#	     sep="\t", quote=F, col.names=NA
#)

top40 <- head(VariableFeatures(etv6_clean2), 40)
top100 <- head(VariableFeatures(etv6_clean2), 100)
```

## Plot variable genes:
```{r }
plot1 <- VariableFeaturePlot(etv6_clean2)
LabelPoints(plot = plot1, points = top40, repel = TRUE)
```

## Standard scaling (no regression)
```{r}
setwd(OUTPUT_PATH)
all.genes <- rownames(etv6_clean2)
etv6_clean2 <- ScaleData(etv6_clean2, features = all.genes, display.progress = TRUE, do.scale=FALSE)
# write.table(etv6_clean2@assays$RNA@scale.data,
#     "log_norm_scaled_filtered_ECFC_MCA_NB_SA.txt",
#     sep="\t", quote=F, col.names=NA
# )

```

## Run PCA
```{r}
etv6_clean2 <- RunPCA(etv6_clean2, features = VariableFeatures(object = etv6_clean2), npcs=40, pcs.print = 5, seed.use=42, rev.pca = FALSE)
```

## Examine PCA results
```{r}
Idents(etv6_clean2) <- etv6_clean2@meta.data$HT0_souporcell_classification
VizDimLoadings(etv6_clean2, dims = 1:2, reduction = "pca")
DimPlot(etv6_clean2, dims=c(1, 2), reduction = "pca", pt.size=1)
DimPlot(etv6_clean2, dims=c(3, 4), reduction = "pca", pt.size=1)
```

## Save PCA coord
```{r}
pca_coordinates <- etv6_clean2@reductions$pca@cell.embeddings
# write.table(pca_coordinates,
# 	   "PCA_coordinates_2020-42-set2_patients.txt",
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Heatmaps:
```{r}
DimHeatmap(etv6_clean2, dims = 1:6, cells = length(colnames(etv6_clean2)), balanced = TRUE)
DimHeatmap(etv6_clean2, dims = 1, cells = length(colnames(etv6_clean2)), balanced = TRUE)
```

## JackStraw and ElbowPlot:
```{r}
etv6_clean2 <- JackStraw(etv6_clean2, num.replicate = 100)
etv6_clean2 <- ScoreJackStraw(etv6_clean2, dims = 1:20)
           
JackStrawPlot(etv6_clean2, dims = 1:15)
ElbowPlot(etv6_clean2)
```

## Perform clustering: 
```{r}
setwd(OUTPUT_PATH)
etv6_clean2 <- FindNeighbors(object = etv6_clean2, k.param = 30, dims = 1:6, compute.SNN = TRUE, prune.SNN = 1/15)

etv6_clean2 <- FindClusters(object= etv6_clean2, modularity.fxn = 1,
             resolution = 0.8, algorithm = 1, n.start = 10, n.iter = 10,
             random.seed = 42, temp.file.location = NULL, edge.file.name = NULL,
             verbose = TRUE)
clusters_res <- etv6_clean2[["RNA_snn_res.0.8"]][,1]
names(clusters_res)=colnames(etv6_clean2)

# write.table(clusters_res,
# 	    "clusters_res_0.8_2020-42-set2_patients.txt",
# 	    sep="\t", quote=F, col.names=NA)
```

## Look at PCA results again, with clusters
```{r}
DimPlot(etv6_clean2, dims=c(1, 2), reduction = "pca", pt.size=1)
DimPlot(etv6_clean2, dims=c(3, 4), reduction = "pca", pt.size=1)
```

## Run UMAP: 
```{r}
setwd(OUTPUT_PATH)
etv6_clean2 <- RunUMAP(etv6_clean2, dims = 1:10, umap.method="uwot", seed.use=10, n.components=2, n.neighbors = 30, spread=1, min.dist=0.2)
DimPlot(etv6_clean2, reduction = "umap", pt.size=0.7, label = TRUE)

# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
UMAP_coord <- etv6_clean2@reductions$umap@cell.embeddings
# write.table(UMAP_coord,
# 	    "UMAP_coordinates_2020-42-set2_patients.txt",
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Run UMAP Cell cycling
```{r}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
 
etv6_clean2 <- CellCycleScoring(etv6_clean2,s.features = s.genes, g2m.features = g2m.genes)
DimPlot(etv6_clean2,group.by="Phase")+ggtitle("UMAP - Cell cycle")
```

## Run tSNE: 
```{r}
etv6_clean2 <- RunTSNE(object= etv6_clean2,  dims = 1:10, dim.embed = 2, seed.use = 42, tsne.method="Rtsne", reduction="pca")
DimPlot(etv6_clean2, reduction = "tsne", pt.size=0.7, label = TRUE)
tSNE_coord <- etv6_clean2@reductions$tsne@cell.embeddings
# write.table(tSNE_coord,
# 	    paste("tSNE_coordinates.",params$experiment,"_merge",".txt", sep=""),
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Top marker genes by clusters and exclude ribosomal genes
```{r}
# ribosomal_genes <- c("RPS17","RPS9","RPS9","RPS9","RPS9","RPS9","RPS9","RPS9","MRPS18B","MRPS18B","MRPS18B","RPS9","RPS18","MRPS36","RPS9","MRPS18B","RPS4Y2","RPS4Y1","RPS25","RPL7A","MRPS6","MRPS18B","RPS18","MRPS18B","RPS18","MRPL57","TBCE","RPS18","MRPS12","MRPS31","RPL3","MRPL23","RPL3L","RPS2","RPL17-C18orf32","RPS4X","RPS13","RPL36","RPS21","RPL21","RPL14","MRPS24","MRPL54","RPS24","RPS15A","RPS6","MRPS25","MRPL36","RPL36AL","MRPS17","MRPL16","MRPS30","FAU","MRPS16","RPS7","RPL23","RPS10","RPL10L","RPL35","RPL17","SRBD1","RPL22","RPL4","MRPS36","RPL37","MRPS2","RPL36A","RPL36A-HNRNPH2","RPL30","MRPL49","RPS19","RPL12","RPS5","RPS27L","RPS29","RPSA","MRPL33","MRPL1","RPS17","RPL10","MRPL13","MRPL19","MRPL15","MRPL41","RPS12","MRPS9","MRPS11","MRPL3","RPL31","RPS3","RPL38","RPL22L1","RPL39","MRPL46","RPL18A","RPL7","RPS3A","RPLP2","RPLP0","RPS16","MRPL23","MRPL22","MRPL18","RPL7L1","MRPL2","MRPS34","MRPS33","RPL28","RPL7A","MRPL4","RPL35A","RPL13","RPL24","RPS20","RPS15","MRPL14","MRPL21","MRPL32","NDUFA7","RPS28","RPL15","RPL18","RPL9","RPL37A","RPS11","MRPS23","MRPL52","MRPL43","MRPS18B","RPLP1","MRPS18A","RSL24D1","RPL27","RPL13A","MRPS14","MRPL47","MRPS35","RPL8","RPL26L1","MRPL11","RPL26","TBCE","RPS27AP5","MRPL34","MRPS12","MRPS5","MRPS18C","RPL29","RPL39L","MRPL28","MRPL51","RPL34","RPL10A","MRPL12","RPL32","RPS14","MRPL35","MRPL17","RPL11","RPS23","RPS9","RPL23A","RPS18","MRPL42","MRPL10","UBA52","DAP3","MRPS7","RPL19","RPS25","RPS26","RPL41","MRPS22","RPL6","RPS8","TBCE","MRPL30","RPS27","RPL27A","MRPS21","MRPL27","MRPL20","MRPL55","MRPL37","RPS27A","MRPL24","RPL5","MRPL9","MRPS15")
# 
# ribosomal_genes_filtered <- ribosomal_genes[!duplicated(ribosomal_genes)]
# all.genes_without_ribosomal = setdiff(all.genes, ribosomal_genes_filtered)
# # length(ribosomal_genes)
# # length(ribosomal_genes_filtered)
# # length(all.genes)
# # length(all.genes_without_ribosomal)
# for(i in 0:(max(as.numeric(levels(etv6_clean2@meta.data$seurat_clusters))))) {
#   assign(paste("cluster", i,".markers", sep = ""), FindMarkers(etv6_clean2, features = all.genes_without_ribosomal, ident.1 = i, test.use = "bimod", only.pos = TRUE))
#     print(paste("cluster", i,".markers", sep = ""))
#     print(head(get(paste("cluster", i,".markers", sep = "")),n=50))
#     write.table(get(paste("cluster", i,".markers", sep = "")),
# 		paste("cluster", i,"markers_2020-42-set2_patients", sep = "_"),
# 		quote=FALSE,row.names = TRUE)
# }
```

## Dimensional reduction plot, with cells colored by a quantitative feature
```{r echo=FALSE}
gene_list <- c("TF","PROCR","CD31","PCAM1","CD34","THBD","ICAM1","MCAM","VCAM1","SELP","CD146","ENG","CDH5","KDR")


for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = etv6_clean2, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 1))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}
```

## Metadata for MCA, NB and SA2
```{r}
setwd(OUTPUT_PATH)
metadata <- as.data.frame(etv6_clean2@meta.data)

#Substitute "RNA_snn_res.0.5" with "RNA_snn_res"
#colnames(metadata)=gsub("RNA_snn_res.[0-9.]+$","RNA_snn_res",colnames(metadata))
metadata$UMAP_1 <- etv6_clean2@reductions$umap@cell.embeddings[,1]
metadata$UMAP_2 <- etv6_clean2@reductions$umap@cell.embeddings[,2]
metadata$tSNE_1 <- etv6_clean2@reductions$tsne@cell.embeddings[,1]
metadata$tSNE_2 <- etv6_clean2@reductions$tsne@cell.embeddings[,2]
metadata$PC1 <- etv6_clean2@reductions[["pca"]]@cell.embeddings[,1]
metadata$PC2 <- etv6_clean2@reductions[["pca"]]@cell.embeddings[,2]

clusters_res <- etv6_clean2[["RNA_snn_res.0.8"]][,1]
names(clusters_res)=colnames(etv6_clean2)

# write.table(metadata,
# 	    "metadata_2020-42-set2_patients.txt",
# 	    sep='\t', quote=F, col.names=NA)
```

## Ggplot for all samples present in the subset dataset and for samples one by one
```{r}
centers=get_cluster_centroids(metadata$UMAP_1,metadata$UMAP_2)
ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.5) + scale_colour_manual(values=color_ordered2) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+
	geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,
		  #nudge_y=-1, nudge_x=-1,
		  size=3)

for(i in sort(unique(metadata$HT0_souporcell_classification))){
  metadata2 <- metadata[which(metadata$HT0_souporcell_classification == i),]
  clusters_res <- metadata2$RNA_snn_res.0.8
  p <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.5) + scale_colour_manual(values=color_ordered2) +       theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(i) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,
		  #nudge_y=-1, nudge_x=-1,
		  size=3)
  plot(p)
}
```

## Save the R object
```{r}
saveRDS(etv6_clean2, file="Patients.rds")
```

## Subset on ctrls
```{r}
clean.subset3 <- subset(x = clean.subset1, idents = c("Ctrl1","Ctrl2"))
```

## Percentage of mitochondrial genes
```{r include=FALSE}

mito.genes <- grep("MT-", rownames(clean.subset3@assays$RNA), value=T)
mito.genes

# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
clean.subset3[["percent.mt"]] <- PercentageFeatureSet(clean.subset3, pattern = "MT-")

# Compute mitochondrial percentage threshold
discard.mito=isOutlier(clean.subset3[["percent.mt"]][,1],type="higher")
mito.threshold=min(clean.subset3[["percent.mt"]][,1][discard.mito])
```

## Filtering on total genes detected and cells with too much mitochondrial genes
```{r}
clean.subset3 <- subset(clean.subset3, subset = (nFeature_RNA > 1000 & percent.mt < mito.threshold))
dim(clean.subset3)
```

## Visualize QC metrics as a violin plot
```{r}
VlnPlot(clean.subset3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3,pt.size=0.1, group.by = "HT0_souporcell_classification")
```

## Normalize data
```{r}
# standard log-normalization
etv6_clean3 <- NormalizeData(clean.subset3, normalization.method = "LogNormalize", scale.factor = 10000)

# write.table(etv6_clean3@assays$RNA@data,
# 	    paste("log_norm_filtered_2020-42-set2_ctrl.txt",sep=""),
# 	    sep="\t", quote=F, col.names=NA)
```

## Some counts after normalization
```{r}
etv6_clean3
table(etv6_clean3$HT0_souporcell_classification)
```

## Choose top 1k variable genes
```{r}
etv6_clean3 <- FindVariableFeatures(object=etv6_clean3, selection.method = "vst", mean.function = ExpMean, dispersion.function = LogVMR, binning.method = "equal_width", num.bin = 20, y.cutoff = 0.5, nfeatures = 1000, verbose=TRUE)
#write.table(icare@assays$RNA@var.features,
#	     paste("Top1000_variable_genes.",experiment,"_",params$sample,".txt", sep=""),
#	     sep="\t", quote=F, col.names=NA
#)

top40 <- head(VariableFeatures(etv6_clean3), 40)
top100 <- head(VariableFeatures(etv6_clean3), 100)
```

## Plot variable genes:
```{r}
plot1 <- VariableFeaturePlot(etv6_clean3)
LabelPoints(plot = plot1, points = top40, repel = TRUE)
```

## Standard scaling (no regression)
```{r}
setwd(OUTPUT_PATH)
all.genes <- rownames(etv6_clean3)
etv6_clean3 <- ScaleData(etv6_clean3, features = all.genes, display.progress = TRUE, do.scale=FALSE)
# write.table(etv6_clean3@assays$RNA@scale.data,
#     "log_norm_scaled_filtered_data_ECFC_SM4_ST_SV2.txt",
#     sep="\t", quote=F, col.names=NA
# )
```

## Run PCA
```{r}
etv6_clean3 <- RunPCA(etv6_clean3, features = VariableFeatures(object = etv6_clean3), npcs=40, pcs.print = 5, seed.use=42, rev.pca = FALSE)
```

## Examine PCA results
```{r}
Idents(etv6_clean3) <- etv6_clean3@meta.data$HT0_souporcell_classification
VizDimLoadings(etv6_clean3, dims = 1:2, reduction = "pca")
DimPlot(etv6_clean3, dims=c(1, 2), reduction = "pca", pt.size=1)
DimPlot(etv6_clean3, dims=c(3, 4), reduction = "pca", pt.size=1)
```

## Save PCA coord
```{r}
pca_coordinates <- etv6_clean3@reductions$pca@cell.embeddings
# write.table(pca_coordinates,
# 	    "PCA_coordinates_2020-42-set2_ctrl.txt",
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Heatmaps:
```{r}
DimHeatmap(etv6_clean3, dims = 1:6, cells = length(colnames(etv6_clean3)), balanced = TRUE)
DimHeatmap(etv6_clean3, dims = 1, cells = length(colnames(etv6_clean3)), balanced = TRUE)
```

## JackStraw and ElbowPlot:
```{r}
etv6_clean3 <- JackStraw(etv6_clean3, num.replicate = 100)
etv6_clean3 <- ScoreJackStraw(etv6_clean3, dims = 1:20)
           
JackStrawPlot(etv6_clean3, dims = 1:15)
           
ElbowPlot(etv6_clean3)
```

## Perform clustering: 
```{r}
setwd(OUTPUT_PATH)
etv6_clean3 <- FindNeighbors(object = etv6_clean3, k.param = 30, dims = 1:6, compute.SNN = TRUE, prune.SNN = 1/15)

etv6_clean3 <- FindClusters(object= etv6_clean3, modularity.fxn = 1,
             resolution = 0.8, algorithm = 1, n.start = 10, n.iter = 10,
             random.seed = 42, temp.file.location = NULL, edge.file.name = NULL,
             verbose = TRUE)
clusters_res <- etv6_clean3[["RNA_snn_res.0.8"]][,1]
names(clusters_res)=colnames(etv6_clean3)

# write.table(clusters_res,
# 	    "clusters_res_0.8_etv6_clean3.txt",
# 	    sep="\t", quote=F, col.names=NA)
```

## Look at PCA results again, with clusters
```{r}
DimPlot(etv6_clean3, dims=c(1, 2), reduction = "pca", pt.size=1)
DimPlot(etv6_clean3, dims=c(3, 4), reduction = "pca", pt.size=1)
```

## Run UMAP: 
```{r}
setwd(OUTPUT_PATH)
etv6_clean3 <- RunUMAP(etv6_clean3, dims = 1:10, umap.method="uwot", seed.use=10, n.components=2, n.neighbors = 30, spread=1, min.dist=0.2)
DimPlot(etv6_clean3, reduction = "umap", pt.size=0.7, label = TRUE)

# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
UMAP_coord <- etv6_clean3@reductions$umap@cell.embeddings
# write.table(UMAP_coord,
# 	    "UMAP_coordinates_2020-42-set2_ctrl.txt",
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Run UMAP Cell cycling
```{r}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
 
etv6_clean3 <- CellCycleScoring(etv6_clean3,s.features = s.genes, g2m.features = g2m.genes)
DimPlot(etv6_clean3,group.by="Phase")+ggtitle("UMAP - Cell cycle")
```

## Run tSNE: 
```{r}
etv6_clean3 <- RunTSNE(object= etv6_clean3,  dims = 1:10, dim.embed = 2, seed.use = 42, tsne.method="Rtsne", reduction="pca")
DimPlot(etv6_clean3, reduction = "tsne", pt.size=0.7, label = TRUE)
tSNE_coord <- etv6_clean3@reductions$tsne@cell.embeddings
# write.table(tSNE_coord,
# 	    paste("tSNE_coordinates.",params$experiment,"_merge",".txt", sep=""),
# 	    sep="\t", quote=F, col.names=NA
# )
```

## Top marker genes by clusters and exclude ribosomal genes
```{r}
# ribosomal_genes <- c("RPS17","RPS9","RPS9","RPS9","RPS9","RPS9","RPS9","RPS9","MRPS18B","MRPS18B","MRPS18B","RPS9","RPS18","MRPS36","RPS9","MRPS18B","RPS4Y2","RPS4Y1","RPS25","RPL7A","MRPS6","MRPS18B","RPS18","MRPS18B","RPS18","MRPL57","TBCE","RPS18","MRPS12","MRPS31","RPL3","MRPL23","RPL3L","RPS2","RPL17-C18orf32","RPS4X","RPS13","RPL36","RPS21","RPL21","RPL14","MRPS24","MRPL54","RPS24","RPS15A","RPS6","MRPS25","MRPL36","RPL36AL","MRPS17","MRPL16","MRPS30","FAU","MRPS16","RPS7","RPL23","RPS10","RPL10L","RPL35","RPL17","SRBD1","RPL22","RPL4","MRPS36","RPL37","MRPS2","RPL36A","RPL36A-HNRNPH2","RPL30","MRPL49","RPS19","RPL12","RPS5","RPS27L","RPS29","RPSA","MRPL33","MRPL1","RPS17","RPL10","MRPL13","MRPL19","MRPL15","MRPL41","RPS12","MRPS9","MRPS11","MRPL3","RPL31","RPS3","RPL38","RPL22L1","RPL39","MRPL46","RPL18A","RPL7","RPS3A","RPLP2","RPLP0","RPS16","MRPL23","MRPL22","MRPL18","RPL7L1","MRPL2","MRPS34","MRPS33","RPL28","RPL7A","MRPL4","RPL35A","RPL13","RPL24","RPS20","RPS15","MRPL14","MRPL21","MRPL32","NDUFA7","RPS28","RPL15","RPL18","RPL9","RPL37A","RPS11","MRPS23","MRPL52","MRPL43","MRPS18B","RPLP1","MRPS18A","RSL24D1","RPL27","RPL13A","MRPS14","MRPL47","MRPS35","RPL8","RPL26L1","MRPL11","RPL26","TBCE","RPS27AP5","MRPL34","MRPS12","MRPS5","MRPS18C","RPL29","RPL39L","MRPL28","MRPL51","RPL34","RPL10A","MRPL12","RPL32","RPS14","MRPL35","MRPL17","RPL11","RPS23","RPS9","RPL23A","RPS18","MRPL42","MRPL10","UBA52","DAP3","MRPS7","RPL19","RPS25","RPS26","RPL41","MRPS22","RPL6","RPS8","TBCE","MRPL30","RPS27","RPL27A","MRPS21","MRPL27","MRPL20","MRPL55","MRPL37","RPS27A","MRPL24","RPL5","MRPL9","MRPS15")
# 
# ribosomal_genes_filtered <- ribosomal_genes[!duplicated(ribosomal_genes)]
# all.genes_without_ribosomal = setdiff(all.genes, ribosomal_genes_filtered)
# # length(ribosomal_genes)
# # length(ribosomal_genes_filtered)
# # length(all.genes)
# # length(all.genes_without_ribosomal)
# for(i in 0:(max(as.numeric(levels(etv6_clean3@meta.data$seurat_clusters))))) {
#   assign(paste("cluster", i,".markers", sep = ""), FindMarkers(etv6_clean3, features = all.genes_without_ribosomal, ident.1 = i, test.use = "bimod", only.pos = TRUE))
#     print(paste("cluster", i,".markers", sep = ""))
#     print(head(get(paste("cluster", i,".markers", sep = "")),n=50))
#     write.table(get(paste("cluster", i,".markers", sep = "")),
# 		paste("cluster", i,"markers_2020-42-set2_ctrl.txt", sep = "_"),
# 		quote=FALSE,row.names = TRUE)
# }
```

## Dimensional reduction plot, with cells colored by a quantitative feature
```{r echo=FALSE}
gene_list <- c("TF","PROCR","CD31","PCAM1","CD34","THBD","ICAM1","MCAM","VCAM1","SELP","CD146","ENG","CDH5","KDR")


for (i in 1:length(gene_list)) {
    tryCatch({
      print(gene_list[i])
      print(FeaturePlot(object = etv6_clean3, features = gene_list[i], reduction = "umap", cols = c("lightgrey", "red1"), pt.size = 1))
    }, error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}

```

## Metadata for SM4, ST and SV2-2
```{r}
setwd(OUTPUT_PATH)
metadata <- as.data.frame(etv6_clean3@meta.data)

#Substitute "RNA_snn_res.0.5" with "RNA_snn_res"
#colnames(metadata)=gsub("RNA_snn_res.[0-9.]+$","RNA_snn_res",colnames(metadata))
metadata$UMAP_1 <- etv6_clean3@reductions$umap@cell.embeddings[,1]
metadata$UMAP_2 <- etv6_clean3@reductions$umap@cell.embeddings[,2]
metadata$tSNE_1 <- etv6_clean3@reductions$tsne@cell.embeddings[,1]
metadata$tSNE_2 <- etv6_clean3@reductions$tsne@cell.embeddings[,2]
metadata$PC1 <- etv6_clean3@reductions[["pca"]]@cell.embeddings[,1]
metadata$PC2 <- etv6_clean3@reductions[["pca"]]@cell.embeddings[,2]

clusters_res <- etv6_clean3[["RNA_snn_res.0.8"]][,1]
names(clusters_res)=colnames(etv6_clean3)

# write.table(metadata,
# 	    "metadata_2020-42-set2_ctrl.txt",
# 	    sep='\t', quote=F, col.names=NA)

```

## Ggplot for all samples present in the subset dataset and for samples one by one
```{r}
centers=get_cluster_centroids(metadata$UMAP_1,metadata$UMAP_2)
ggplot(metadata, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.5) + scale_colour_manual(values=color_ordered2) + theme(panel.background = element_rect(fill = 'white', colour = 'black'))+
	geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,
		  #nudge_y=-1, nudge_x=-1,
		  size=3)

for(i in sort(unique(metadata$HT0_souporcell_classification))){
  metadata2 <- metadata[which(metadata$HT0_souporcell_classification == i),]
  clusters_res <- metadata2$RNA_snn_res.0.8
  p <- ggplot(metadata2, aes(x=UMAP_1, y=UMAP_2)) + geom_point(aes(colour=factor(clusters_res)), size=0.5) + scale_colour_manual(values=color_ordered2) +       theme(panel.background = element_rect(fill = 'white', colour = 'black'))+ ggtitle(i) + geom_text(aes(x=x_center,y=y_center,label=Cluster),data=centers,
		  #nudge_y=-1, nudge_x=-1,
		  size=3)
  plot(p)
}
```

## Save the R object
```{r}
saveRDS(etv6_clean3, file="Ctrl.rds")
```